gorabbitmq_test.go

```diff
@@ -0,0 +1,1669 @@
+package gorabbitmq_test
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"os"
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/Clarilab/gorabbitmq/v5"
+	amqp "github.com/rabbitmq/amqp091-go"
+)
+
+const (
+	logLevel = slog.LevelError
+)
+
+type testData struct {
+	Name    string `json:"name"`
+	Age     int    `json:"age"`
+	City    string `json:"city"`
+	Country string `json:"country"`
+}
+
+func Test_Integration_PublishToExchange(t *testing.T) { //nolint:maintidx // table testing
+	t.Parallel()
+
+	stringMessage := "test-message"
+	bytesMessage := []byte(stringMessage)
+
+	jsonMessage := testData{
+		Name:    "Name",
+		Age:     157,
+		City:    "City",
+		Country: "Country",
+	}
+
+	type testParams struct {
+		exchangeName string
+		queueName    string
+		routingKey   string
+	}
+
+	tests := map[string]struct {
+		testParams      *testParams
+		connector       *gorabbitmq.Connector
+		deliveryHandler func(any, int, chan struct{}) gorabbitmq.HandlerFunc
+		preConsumer     func(*gorabbitmq.Connector, gorabbitmq.HandlerFunc, *testParams) (*gorabbitmq.Consumer, error)
+		getConsumer     func(*gorabbitmq.Connector, gorabbitmq.HandlerFunc, *testParams) (*gorabbitmq.Consumer, error)
+		getPublisher    func(*gorabbitmq.Connector, *testParams) (*gorabbitmq.Publisher, error)
+		publish         func(*testParams) error
+		message         any
+		doneChan        chan struct{}
+	}{
+		"publish to exchange / consume with Ack": {
+			testParams: &testParams{
+				exchangeName: stringGen(),
+				queueName:    stringGen(),
+				routingKey:   stringGen(),
+			},
+			connector: func() *gorabbitmq.Connector {
+				connector, err := gorabbitmq.NewConnector(&gorabbitmq.ConnectionSettings{
+					UserName: "guest",
+					Password: "guest",
+					Host:     "localhost",
+					Port:     5672,
+				}, gorabbitmq.WithConnectorOptionLogHandler(
+					slog.NewTextHandler(
+						os.Stdout,
+						&slog.HandlerOptions{
+							Level: logLevel,
+						},
+					),
+				),
+					gorabbitmq.WithConnectorOptionConnectionName("test-connection"),
+				)
+				requireNoError(t, err)
+
+				return connector
+			}(),
```

Comment:
does this snippet ever change in all of those test cases? seems useless to declare it 20 times 😄